### ISSUE TITLE
Indicate "diff" for `assertEquals()` failures

### ISSUE THREAD

**Issue by delanym** [junit-team/junit-framework#3192]
When I compare two strings
`Assertions.assertEquals(expected, fb.toString());`
there's no indication from the error what the difference is if its only whitespace. 

So I need to compare the byte arrays instead.
`Assertions.assertArrayEquals(expected.getBytes(), fb.toString().getBytes());`

If there's a failure I'd like to see the string version (to make sure its roughly the correct string) AND the byte values, but I don't want to waste time comparing both of them to satisfy the test condition.

I guess there are many way to solve this. Maybe its possible to compare the values like assertEquals, and if the test fails, output as usual, but append a diff. "assertNoDiff", "assertEqualsDiff" etc.

**Comment by astocker-scottlogic** [comment]
Am I okay to have a crack at this, assuming no one else has picked it up?

**Comment by igmtz** [comment]
Hey, can I also work on this issue?

**Comment by delanym** [comment]
@igmtz please have a go at it

**Comment by marcphilipp** [comment]
@igmtz Before this can be started we need to agree an an approach. What did you have in mind? 🙂

**Comment by astocker-scottlogic** [comment]
@marcphilipp How about wrapping the empty strings in quotation marks and then displaying the two results?

**Comment by astocker-scottlogic** [comment]
Hi @marcphilipp, do you think the approach I wrote above would be suitable? Wrapping the strings in quotes - I'm worried about starting the work if it's not the best idea

**Comment by delanym** [comment]
@astocker-scottlogic they are not empty strings: they're strings that _differ_ by whitespace only. The method would also be useful when glyphs have a similar appearance but are in fact different, i.e. different code points.

I've been in a situation where a colleague thought they were looking at the same string. But on my machine, using a different font, the difference was evident.

**Comment by marcphilipp** [comment]
Some research:

JUnit 4 highlights the differences between two strings with square brackets (see `ComparisonFailure`). For example:

```
org.junit.ComparisonFailure: message expected:<a[	]c> but was:<a[    ]c>
```

Spock does this:

```
Condition not satisfied:
"a    c" == "a\tc"
         |
         false
         4 differences (33% similarity)
         a( ~   )c
         a(\t---)c
```

JUnit Jupiter currently does this:

```
org.opentest4j.AssertionFailedError: expected: <a    c> but was: <a	c>
```

We could extend the above message in case we're comparing Strings to show:

```
org.opentest4j.AssertionFailedError: expected: <a    c> but was: <a	c>
diff: a[    ]c
      a[\t]c
```

That would replace all control characters with their symbol (`\t`, `\n`, ...), if known, or their unicode representation (e.g. `\u0009`).

@delanym WDYT?

**Comment by delanym** [comment]
Looks good! Would the diff come in color? In which case it could be a single line (red square brackets green curly)

    org.opentest4j.AssertionFailedError: expected: <a    c> but was: <a	c>
    diff: a[    ]{\t}c

Git adds plus/minus

    org.opentest4j.AssertionFailedError: expected: <a    c> but was: <a	c>
    diff: a[-    -]{+\t+}c

**Comment by marcphilipp** [comment]
I don't think we can do color in the exception message. We could maybe do this in the ConsoleLauncher. There's a similar issue to output a diff there: #3139.

**Comment by KarthikRv93** [comment]
Can we proceed with this issue using the proposed solution mentioned above?
https://github.com/junit-team/junit-framework/issues/3192#issuecomment-1635720881
PR: https://github.com/junit-team/junit-framework/pull/4787

**Comment by marcphilipp** [comment]
I have some doubts whether the exception message is the right place to include a diff. IDEs such as Eclipse and IntelliJ IDEA already display a diff based on `expected` and `actual` in `AssertionFailedError`. The ConsoleLauncher could also do that (see #3139). And the Open Test Reporting HTML report could do this as well.

@KarthikRv93 What's your use case where you currently don't get a diff? CI builds?
